crates/ansi-to-html/src/test_utils.rs

```diff
@@ -0,0 +1,327 @@
+                match tag_kind {
+                    TagKind::StartTag => interpreter.apply_start_style(raw_style),
+                    TagKind::EndTag => interpreter.apply_end_style(raw_style),
+                }
+            }
+            Token::CharacterTokens(s) => {
+                let mut interpreter = self.0.borrow_mut();
+                interpreter.push_pending_text(&s);
+            }
+            Token::NullCharacterToken | Token::ParseError(_) | Token::EOFToken => {}
+            unknown => panic!("Missing implementation for {unknown:#?}"),
+        }
+
+        TokenSinkResult::Continue
+    }
+}
+
+#[derive(Default)]
+struct Inner {
+    state: State,
+    output: Vec<StylizedText>,
+}
+
+impl Inner {
+    fn apply_start_style(&mut self, style: RawStyle) {
+        self.emit_pending_text();
+
+        if let Some(last_last) = self.state.last_style.replace(style) {
+            self.state.raw_styles.push(last_last);
+        }
+    }
+
+    fn apply_end_style(&mut self, style: RawStyle) {
+        self.emit_pending_text();
+
+        let mut start_tag = match self.state.last_style.take() {
+            // We just pushed an opening tag without pushing text, so this is an empty span
+            Some(style) => {
+                let mut styles = Styles::new(&self.state.raw_styles);
+                styles = styles.apply(style.clone());
+                self.output.push(StylizedText::empty(styles));
+                style
+            }
+            None => self.state.raw_styles.pop().unwrap(),
+        };
+
+        // The end tag won't have any of the attrs that the start had
+        if let RawStyle::Span(attrs) = &mut start_tag {
+            let _ = mem::take(attrs);
+        }
+
+        assert_eq!(start_tag, style, "Start tag should have a matching end");
+    }
+
+    fn push_pending_text(&mut self, s: &str) {
+        if let Some(pending) = self.state.last_style.take() {
+            self.state.raw_styles.push(pending);
+        }
+
+        self.state.text = match self.state.text.take() {
+            Some(mut text) => {
+                text.push_str(s);
+                Some(text)
+            }
+            None => Some(s.to_owned()),
+        }
+    }
+
+    fn emit_pending_text(&mut self) {
+        // `last_state` gets cleared out when text gets pushed, so it being `Some` means there's no
+        // pending text
+        if self.state.last_style.is_some() {
+            return;
+        }
+
+        if let Some(text) = self.state.text.take() {
+            let styles = Styles::new(&self.state.raw_styles);
+            let stylized_text = StylizedText::new(styles, text);
+            self.output.push(stylized_text);
+        }
+    }
+}
+
+#[derive(Default)]
+struct State {
+    /// The style that was just pushed on before any text
+    ///
+    /// We buffer the styles that we push, so that we can identify and retain empty pairs of
+    /// elements as empty text
+    last_style: Option<RawStyle>,
```

Comment:
Somewhat awkwardly we go through extra effort to retain empty spans here just to discard them later in `normalize_output()`, but ideally in the future the optimized output won't have empty spans to begin with, so we can avoid filtering them out to ensure that too (heavy emphasis on ideally of course 🤞)

Reply:
Oh and I just realized that now that `text` is an `Option<String>` which can indicate the absence of pending text. That means `last_state` can be folded back into `raw_styles`. That should clean up the implementation a good bit